lib/glific/third_party/sheets/sheet_data.ex

```diff
@@ -0,0 +1,139 @@
+defmodule Glific.Sheets.SheetData do
+  @moduledoc """
+  The minimal wrapper for the base Sheet structure
+  """
+  use Ecto.Schema
+  import Ecto.Changeset
+  import Ecto.Query
+
+  alias __MODULE__
+
+  alias Glific.{
+    Partners.Organization,
+    Repo,
+    Sheets.ApiClient,
+    Sheets.Sheet
+  }
+
+  @required_fields [
+    :key,
+    :data,
+    :sheet_id,
+    :organization_id,
+    :synced_at
+  ]
+
+  @type t() :: %__MODULE__{
+          __meta__: Ecto.Schema.Metadata.t(),
+          id: non_neg_integer | nil,
+          key: String.t() | nil,
+          data: map() | nil,
+          synced_at: :utc_datetime | nil,
+          organization_id: non_neg_integer | nil,
+          organization: Organization.t() | Ecto.Association.NotLoaded.t() | nil,
+          sheet_id: non_neg_integer | nil,
+          sheet: Sheet.t() | Ecto.Association.NotLoaded.t() | nil,
+          inserted_at: :utc_datetime | nil,
+          updated_at: :utc_datetime | nil
+        }
+
+  schema "sheets_data" do
+    field(:key, :string)
+    field(:data, :map, default: %{})
+    field(:synced_at, :utc_datetime)
+
+    belongs_to(:sheet, Sheet)
+    belongs_to(:organization, Organization)
+
+    timestamps(type: :utc_datetime_usec)
+  end
+
+  @doc """
+  Standard changeset pattern we use for all data types
+  """
+  @spec changeset(SheetData.t(), map()) :: Ecto.Changeset.t()
+  def changeset(sheet_data, attrs) do
+    sheet_data
+    |> cast(attrs, @required_fields)
+    |> validate_required(@required_fields)
+    |> unique_constraint([:key, :sheet_id, :organization_id])
+  end
+
+  @doc """
+  Parses a sheet
+  """
+  @spec parse_sheet_data(map(), Sheet.t()) :: :ok
+  def parse_sheet_data(attrs, sheet) do
+    ApiClient.get_csv_content(url: attrs.url)
+    |> Enum.each(fn {_, row} ->
+      %{
+        key: row["Key"],
+        data: clean_row_values(row),
+        sheet_id: sheet.id,
+        synced_at: sheet.synced_at,
+        organization_id: attrs.organization_id
+      }
+      |> upsert_sheet_data()
+    end)
+
+    clean_unsynced_sheet_data(sheet)
+    :ok
+  end
+
+  @spec clean_row_values(map()) :: map()
+  defp clean_row_values(row) do
+    Enum.reduce(row, %{}, fn {key, value}, acc ->
+      key = key |> String.downcase() |> String.replace(" ", "_")
```

Comment:
Let's use elixir's inbuilt functions for this. 
https://hexdocs.pm/elixir/1.13.1/Macro.html#underscore/1-examples

Reply:
@pankaj-ag Macro functions fail in case
`"foo bar" |> String.replace(" ", "") |> Macro.underscore()` gives  `foobar`
`"Foo bar" |> String.replace(" ", "") |> Macro.underscore()` gives  `foobar`